smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/ResourceOperationInputOutputValidator.java

```diff
@@ -0,0 +1,166 @@
+        });
+
+        resource.getDelete().flatMap(model::getShape).flatMap(Shape::asOperationShape).ifPresent(operation -> {
+            propertiesInOperations.addAll(getAllOperationProperties(model, operation));
+            validateTopLevelMembersToProperties(model, resource, operation, true, "delete", events);
+        });
+
+        resource.getList().flatMap(model::getShape).flatMap(Shape::asOperationShape).ifPresent(operation -> {
+            propertiesInOperations.addAll(getAllOperationProperties(model, operation));
+            validateTopLevelMembersToProperties(model, resource, operation, true, "list", events);
+        });
+
+        Set<String> definedProperties = resource.getProperties().keySet();
+        definedProperties.retainAll(propertiesInOperations);
+        for (String propertyNotInLifecycleOp : definedProperties) {
+            events.add(error(resource, String.format("Resource shape's `%s` property which is not an input or output"
+                     + " for any lifecycle operation.", propertyNotInLifecycleOp)));
+        }
+
+        return events;
+    }
+
+    private List<String> getAllOperationProperties(Model model, OperationShape operation) {
+        List<String> properties = new ArrayList<>();
+
+        operation.getInput().flatMap(model::getShape).flatMap(Shape::asStructureShape).ifPresent(structure -> {
+            for (MemberShape member : structure.members()) {
+                properties.add(getPropertyNameFromMember(member));
+            }
+        });
+        operation.getOutput().flatMap(model::getShape).flatMap(Shape::asStructureShape).ifPresent(structure -> {
+            for (MemberShape member : structure.members()) {
+                properties.add(getPropertyNameFromMember(member));
+            }
+        });
+        return properties;
+    }
+
+    private String getPropertyNameFromMember(MemberShape member) {
+        return member.getTrait(PropertyTrait.class)
+                    .flatMap(trait -> trait.getName()).orElse(member.getMemberName());
+    }
+
+    private void validateTopLevelMembersToProperties(
+        Model model,
+        ResourceShape resource,
+        OperationShape operation,
+        boolean areOutputProperties,
+        String lifecycleOperationName,
+        List<ValidationEvent> events
+    ) {
+        Optional<ShapeId> topLevelShapeId =
+                areOutputProperties
+                ? operation.getOutput()
+                : operation.getInput();
+        Map<String, ShapeId> properties = resource.getProperties();
+        Map<String, Set<MemberShape>> propertyToMemberMappings = new TreeMap<>();
+
+        Optional<StructureShape> topLevelShape = topLevelShapeId
+                        .flatMap(model::getShape).flatMap(Shape::asStructureShape);
+        topLevelShape.ifPresent(shape -> {
+            for (MemberShape member : shape.members()) {
+                if (!member.hasTrait(TransientTrait.class)) {
```

Comment:
This should handle traits that are marked with TransientTrait as well.

Reply:
Encapsulated the necessary logic in a method.